File: 200-number-of-islands/number-of-islands.cpp
class Solution {
public:
    int n,m;
    void dfs(vector<vector<char>>& grid,vector<vector<int>>& visited,int i,int j)
    {
        if(i>=n || j >=m || i<0 || j<0)
        {
            return;
        }
        if(i+1< n && visited[i+1][j]==-1 && grid[i+1][j]=='1')
        {
            visited[i+1][j]=1;
            dfs(grid,visited,i+1,j);
        }
        if(j-1>=0 && visited[i][j-1]==-1 && grid[i][j-1]=='1')
        {
            visited[i][j-1]=1;
            dfs(grid,visited,i,j-1);
        }
        if(i-1>=0 && visited[i-1][j]==-1 && grid[i-1][j]=='1')
        {
            visited[i-1][j]=1;
            dfs(grid,visited,i-1,j);
        }
        if(j+1<m && visited[i][j+1]==-1 && grid[i][j+1]=='1')
        {
            visited[i][j+1]=1;
            dfs(grid,visited,i,j+1);
        }
    }
    int numIslands(vector<vector<char>>& grid) {
         n= grid.size(),m=grid[0].size();
         int c=0;
        vector<vector<int>> visited(n+1,vector<int>(m+1,-1));
        for(int i=0;i<n;i++)
        {
            for(int j=0;j<m;j++)
            {
                if(grid[i][j]=='1' && visited[i][j]==-1)
                {   visited[i][j]=1;
                    c++;
                    dfs(grid,visited,i,j);
                }
            }
        }
        return c;
    }
};

File: 200-number-of-islands/number-of-islands.java
class Solution {
    void dfs(char[][] grid,int i,int j,int[][] visited,int n,int m)
    {
        visited[i][j]=1;
        if(i+1<n && visited[i+1][j]==0 && grid[i+1][j]=='1')
        {
            dfs(grid,i+1,j,visited,n,m);
        }
        if(j+1<m && visited[i][j+1]==0 && grid[i][j+1]=='1')
        {
            dfs(grid,i,j+1,visited,n,m);
        }
        if(i-1>=0 && visited[i-1][j]==0 && grid[i-1][j]=='1')
        {
            dfs(grid,i-1,j,visited,n,m);
        }
        if(j-1>=0 && visited[i][j-1]==0 && grid[i][j-1]=='1')
        {
            dfs(grid,i,j-1,visited,n,m);
        }

    }
    public int numIslands(char[][] grid) {
        int n=grid.length;
        int m=grid[0].length;
        int[][] visited = new int[n][m];
        for(int i=0;i<n;i++)
        {
            for(int j=0;j<m;j++)
            {
                visited[i][j]=0;
            }
        }
        int c=0;
        for(int i=0;i<n;i++)
        {
            for(int j=0;j<m;j++)
            {
                if(visited[i][j]==0 && grid[i][j]=='1')
                {
                    c++;
                    dfs(grid,i,j,visited,n,m);
                }
            }
        }    
        return c;
    }
}